File: server/process.ts
// Process manager for live-capture.exe instances.
//
// Each capture stream is a child process that writes binary IPC messages to
// stdout.  This module spawns them, wires stdout through the ProtocolParser
// into a StreamBuffer, and handles lifecycle events.

import type { Subprocess } from "bun";

import { StreamBuffer } from "./buffer";
import { captureExePath, frameBufferCapacity } from "./common";
import { ProtocolParser } from "./protocol";

// ── Types ────────────────────────────────────────────────────────────────────

export interface CaptureStream {
    id: string;
    hwnd: string;
    status: "starting" | "running" | "stopped";
    buffer: StreamBuffer;
    /// Monotonically increasing counter — bumped each time the underlying
    /// capture process is replaced (via replaceStream / replaceCropStream).
    /// The frontend uses this to detect stream replacement and reinitialize
    /// its decoder.
    generation: number;
    /// Bun child process handle.  Null after the process has exited and been
    /// cleaned up by destroyStream().
    process: Subprocess | null;
}

// ── Registry ─────────────────────────────────────────────────────────────────

const streams = new Map<string, CaptureStream>();

export function getStream(id: string): CaptureStream | undefined {
    return streams.get(id);
}

export function listStreams(): CaptureStream[] {
    return [...streams.values()];
}

// ── Enumerate windows ────────────────────────────────────────────────────────

/// Spawn `live-capture.exe --enumerate-windows` and return the JSON array
/// of capturable windows.  This is a one-shot process, not a long-running
/// capture.
export async function enumerateWindows(): Promise<unknown[]> {
    const proc = Bun.spawn([captureExePath, "--enumerate-windows"], {
        stdout: "pipe",
        stderr: "pipe",
    });

    const stdout = await new Response(proc.stdout).text();
    await proc.exited;

    return JSON.parse(stdout) as unknown[];
}

// ── Create / destroy streams ─────────────────────────────────────────────────

/// Spawn a live-capture.exe child process and wire its stdout/stderr to the
/// given `CaptureStream`.  The stream must already be registered in the Map.
///
/// Factored out of `spawnCapture` so both the create and replace paths can
/// share the same wiring logic without duplication.
function spawnAndWire(stream: CaptureStream, args: string[], label: string): void {
    const { id } = stream;
    const proc = Bun.spawn(args, { stdout: "pipe", stderr: "pipe" });
    stream.process = proc;

    // Wire stdout → ProtocolParser → StreamBuffer
    const parser = new ProtocolParser((msg) => {
        switch (msg.type) {
            case "codec_params":
                stream.buffer.setCodecParams(msg.params);
                if (stream.status === "starting") {
                    stream.status = "running";
                    console.log(`[stream:${id}] running (codec params received)`);
                }
                break;
            case "frame":
                stream.buffer.pushFrame(msg.frame);
                break;
            case "error":
                console.error(`[stream:${id}] capture error: ${msg.message}`);
                break;
        }
    });

    pipeStdout(id, proc, parser);
    pipeStderr(id, proc);

    proc.exited.then((code) => {
        console.log(`[stream:${id}] process exited with code ${code}`);
        stream.status = "stopped";
    });

    console.log(`[stream:${id}] spawned (${label})`);
}

/// Create a brand-new stream with a random ID and register it.
/// Used by the manual create endpoints (live-control) and as the internal
/// fallback when replaceStream / replaceCropStream targets a non-existent ID.
function spawnCapture(id: string, hwnd: string, args: string[], label: string): CaptureStream {
    const buffer = new StreamBuffer(frameBufferCapacity);

    const stream: CaptureStream = {
        id, hwnd,
        status: "starting",
        buffer,
        generation: 1,
        process: null,
    };

    streams.set(id, stream);
    spawnAndWire(stream, args, label);
    return stream;
}

/// Spawn a resample-mode capture: scales the full window to `width x height`.
export function createStream(hwnd: string, width: number, height: number): CaptureStream {
    const id = crypto.randomUUID().slice(0, 8);
    return spawnCapture(id, hwnd,
        [captureExePath, "--hwnd", hwnd, "--width", String(width), "--height", String(height)],
        `hwnd=${hwnd}, resample ${width}x${height}`);
}

/// Spawn a crop-mode capture: extracts a subrect at native resolution.
///
/// `cropWidth`/`cropHeight` are either a pixel count (multiple of 16) or
/// `"full"` for the source dimension.  `cropAlign` controls placement of the
/// crop rect within the source window (e.g. "bottom", "center").
export function createCropStream(
    hwnd: string,
    cropWidth: "full" | number,
    cropHeight: "full" | number,
    cropAlign: string): CaptureStream {
    const id = crypto.randomUUID().slice(0, 8);
    return spawnCapture(id, hwnd,
        [captureExePath, "--hwnd", hwnd,
            "--crop-width", String(cropWidth),
            "--crop-height", String(cropHeight),
            "--crop-align", cropAlign],
        `hwnd=${hwnd}, crop ${cropWidth}x${cropHeight} ${cropAlign}`);
}

/// Replace the capture process behind a well-known stream ID.
///
/// If a stream with `id` already exists, the old process is killed, the
/// buffer is reset, and the generation counter is bumped — the Map entry is
/// never removed, so `getStream(id)` never returns `undefined` during the
/// transition.  If no stream with `id` exists, one is created (idempotent).
export function replaceStream(id: string, hwnd: string, width: number, height: number): CaptureStream {
    const args = [captureExePath, "--hwnd", hwnd, "--width", String(width), "--height", String(height)];
    const label = `hwnd=${hwnd}, resample ${width}x${height}`;

    const existing = streams.get(id);
    if (!existing) return spawnCapture(id, hwnd, args, label);

    // Kill old process, reset buffer, bump generation.
    if (existing.process) {
        existing.process.kill();
        existing.process = null;
    }
    existing.hwnd = hwnd;
    existing.status = "starting";
    existing.buffer.reset();
    existing.generation++;

    spawnAndWire(existing, args, label);
    console.log(`[stream:${id}] replaced (gen ${existing.generation})`);
    return existing;
}

/// Replace a crop-mode stream behind a well-known stream ID.
/// Same semantics as `replaceStream` but for crop mode.
export function replaceCropStream(
    id: string,
    hwnd: string,
    cropWidth: "full" | number,
    cropHeight: "full" | number,
    cropAlign: string): CaptureStream {
    const args = [
        captureExePath, "--hwnd", hwnd,
        "--crop-width", String(cropWidth),
        "--crop-height", String(cropHeight),
        "--crop-align", cropAlign];
    const label = `hwnd=${hwnd}, crop ${cropWidth}x${cropHeight} ${cropAlign}`;

    const existing = streams.get(id);
    if (!existing) return spawnCapture(id, hwnd, args, label);

    if (existing.process) {
        existing.process.kill();
        existing.process = null;
    }
    existing.hwnd = hwnd;
    existing.status = "starting";
    existing.buffer.reset();
    existing.generation++;

    spawnAndWire(existing, args, label);
    console.log(`[stream:${id}] replaced (gen ${existing.generation})`);
    return existing;
}

/// Kill the child process and remove the stream from the registry.
export function destroyStream(id: string): void {
    const stream = streams.get(id);
    if (!stream) return;

    if (stream.process) {
        stream.process.kill();
        stream.process = null;
    }

    streams.delete(id);
    console.log(`[stream:${id}] destroyed`);
}

/// Kill all child processes.  Called on server shutdown.
export function destroyAll(): void {
    for (const [id] of streams) {
        destroyStream(id);
    }
}

// ── Internal helpers ─────────────────────────────────────────────────────────

/// Read stdout chunks from the child process and feed them to the parser.
async function pipeStdout(id: string, proc: Subprocess, parser: ProtocolParser): Promise<void> {
    try {
        const reader = proc.stdout.getReader();
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            parser.feed(value);
        }
    } catch (e) {
        // Expected when the process is killed — the stream closes.
        console.log(`[stream:${id}] stdout closed`);
    }
}

/// Forward stderr lines with a prefix for easy identification in the console.
async function pipeStderr(id: string, proc: Subprocess): Promise<void> {
    try {
        const reader = proc.stderr.getReader();
        const decoder = new TextDecoder();
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            process.stderr.write(`[capture:${id}] ${decoder.decode(value)}`);
        }
    } catch {
        // Expected on process kill.
    }
}

File: live-server/src/video/process.rs
//! Process manager for `live-video.exe` instances.
//!
//! Spawns child processes, reads their stdout via `live_video::read_message()`,
//! and pushes parsed frames into the stream's `VideoBuffer`.

use crate::video::buffer::StreamBuffer;

use live_video::Message;

use std::collections::HashMap;
use std::io::BufReader;
use std::process::{Child, Command, Stdio};
use std::sync::Arc;

use job_object::JobObject;
use tokio::sync::RwLock;
use tokio::task::JoinHandle;

use crate::constant::FRAME_BUFFER_CAPACITY;

// ── Types ────────────────────────────────────────────────────────────────────

/// Status of a capture stream.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "lowercase")]
pub enum StreamStatus {
    Starting,
    Running,
    Stopped,
}

/// A single capture stream backed by a `live-video.exe` child process.
pub struct VideoStream {
    pub id: String,
    pub hwnd: String,
    pub status: StreamStatus,
    pub buffer: StreamBuffer,
    /// Bumped each time the underlying capture process is replaced.
    pub generation: u32,
    /// Handle to the child process.
    pub child: Option<Child>,
    /// Abort handle for the stdout reader task.
    pub reader_handle: Option<JoinHandle<()>>,
}

impl VideoStream {
    /// Kill the child process and abort the reader task.
    fn kill(&mut self) {
        if let Some(mut child) = self.child.take() {
            let _ = child.kill();
            let _ = child.wait();
        }
        if let Some(handle) = self.reader_handle.take() {
            handle.abort();
        }
        self.status = StreamStatus::Stopped;
    }
}

impl Drop for VideoStream {
    fn drop(&mut self) { self.kill(); }
}

// ── Stream Registry ──────────────────────────────────────────────────────────

/// Registry of all active capture streams.
///
/// The `RwLock` is held briefly for reads (frame polling at 60fps) and writes
/// (stream create/destroy, frame push).  Individual stream buffers are accessed
/// within the lock — the granularity is the entire registry, not per-stream,
/// keeping the implementation simple for the expected 1-3 concurrent streams.
pub struct StreamRegistry {
    pub streams: HashMap<String, VideoStream>,
    /// Path to the `live-video.exe` binary.
    pub exe_path: String,
    /// Job object — assigned children are killed when the server exits.
    pub job: Arc<JobObject>,
}

impl StreamRegistry {
    pub fn new(exe_path: String, job: Arc<JobObject>) -> Self {
        Self {
            streams: HashMap::new(),
            exe_path,
            job,
        }
    }

    /// List all streams as serializable info structs.
    pub fn list(&self) -> Vec<StreamInfo> {
        self.streams.values().map(|s| StreamInfo {
            id: s.id.clone(),
            hwnd: s.hwnd.clone(),
            status: s.status,
            generation: s.generation,
        }).collect()
    }

    /// Create a resample-mode stream with a random ID.
    pub fn create_stream(
        &mut self,
        hwnd: &str,
        width: u32,
        height: u32,
        registry: &std::sync::Arc<RwLock<Self>>,
    ) -> String {
        let id = short_id();
        let args = vec![
            self.exe_path.clone(),
            "--hwnd".into(), hwnd.into(),
            "--width".into(), width.to_string(),
            "--height".into(), height.to_string(),
        ];
        self.spawn_capture(&id, hwnd, &args, registry);
        id
    }

    /// Create a crop-mode stream with a random ID.
    pub fn create_crop_stream(
        &mut self,
        hwnd: &str,
        crop: &CropParams,
        fps: Option<u32>,
        registry: &std::sync::Arc<RwLock<Self>>,
    ) -> String {
        let id = short_id();
        let mut args = vec![
            self.exe_path.clone(),
            "--hwnd".into(), hwnd.into(),
            "--crop-min-x".into(), crop.min_x.to_string(),
            "--crop-min-y".into(), crop.min_y.to_string(),
            "--crop-max-x".into(), crop.max_x.to_string(),
            "--crop-max-y".into(), crop.max_y.to_string(),
        ];
        if let Some(fps) = fps {
            args.push("--fps".into());
            args.push(fps.to_string());
        }
        self.spawn_capture(&id, hwnd, &args, registry);
        id
    }

    /// Replace a well-known stream in-place (kill old process, bump generation).
    /// Creates a new stream if none exists with this ID.
    pub fn replace_stream(
        &mut self,
        id: &str,
        hwnd: &str,
        width: u32,
        height: u32,
        registry: &std::sync::Arc<RwLock<Self>>,
    ) {
        let args = vec![
            self.exe_path.clone(),
            "--hwnd".into(), hwnd.into(),
            "--width".into(), width.to_string(),
            "--height".into(), height.to_string(),
        ];
        self.replace_or_create(id, hwnd, &args, registry);
    }

    /// Replace a well-known crop stream in-place.
    pub fn replace_crop_stream(
        &mut self,
        id: &str,
        hwnd: &str,
        crop: &CropParams,
        fps: Option<u32>,
        registry: &std::sync::Arc<RwLock<Self>>,
    ) {
        let mut args = vec![
            self.exe_path.clone(),
            "--hwnd".into(), hwnd.into(),
            "--crop-min-x".into(), crop.min_x.to_string(),
            "--crop-min-y".into(), crop.min_y.to_string(),
            "--crop-max-x".into(), crop.max_x.to_string(),
            "--crop-max-y".into(), crop.max_y.to_string(),
        ];
        if let Some(fps) = fps {
            args.push("--fps".into());
            args.push(fps.to_string());
        }
        self.replace_or_create(id, hwnd, &args, registry);
    }

    /// Destroy a stream by ID.
    pub fn destroy_stream(&mut self, id: &str) {
        if let Some(mut stream) = self.streams.remove(id) {
            stream.kill();
            log::info!("[{id}] destroyed");
        }
    }

    /// Kill all child processes.  Called on server shutdown.
    pub fn destroy_all(&mut self) {
        let ids: Vec<_> = self.streams.keys().cloned().collect();
        for id in ids {
            self.destroy_stream(&id);
        }
    }

    // ── Internal ─────────────────────────────────────────────────────────

    fn replace_or_create(
        &mut self,
        id: &str,
        hwnd: &str,
        args: &[String],
        registry: &std::sync::Arc<RwLock<Self>>,
    ) {
        if let Some(stream) = self.streams.get_mut(id) {
            stream.kill();
            hwnd.clone_into(&mut stream.hwnd);
            stream.status = StreamStatus::Starting;
            stream.buffer.reset();
            stream.generation += 1;
            let generation = stream.generation;

            let (child, reader_handle) = spawn_and_wire(id, args, &self.job, registry);
            stream.child = Some(child);
            stream.reader_handle = Some(reader_handle);

            log::info!("[{id}] replaced (gen {generation})");
        } else {
            self.spawn_capture(id, hwnd, args, registry);
        }
    }

    fn spawn_capture(
        &mut self,
        id: &str,
        hwnd: &str,
        args: &[String],
        registry: &std::sync::Arc<RwLock<Self>>,
    ) {
        let (child, reader_handle) = spawn_and_wire(id, args, &self.job, registry);

        let stream = VideoStream {
            id: id.to_owned(),
            hwnd: hwnd.to_owned(),
            status: StreamStatus::Starting,
            buffer: StreamBuffer::new(FRAME_BUFFER_CAPACITY),
            generation: 1,
            child: Some(child),
            reader_handle: Some(reader_handle),
        };

        self.streams.insert(id.to_owned(), stream);
        log::info!("[{id}] spawned");
    }
}

/// Crop region parameters for crop-mode capture.
pub struct CropParams {
    pub min_x: u32,
    pub min_y: u32,
    pub max_x: u32,
    pub max_y: u32,
}

/// Serializable stream info for the list endpoint.
#[derive(serde::Serialize)]
pub struct StreamInfo {
    pub id: String,
    pub hwnd: String,
    pub status: StreamStatus,
    pub generation: u32,
}

// ── Child Process + Reader Task ──────────────────────────────────────────────

/// Spawn a `live-video.exe` child process and start a tokio task that reads
/// its stdout via `live_video::read_message()`.
///
/// Returns the child process handle and the reader task's join handle.
fn spawn_and_wire(
    id: &str,
    args: &[String],
    job: &JobObject,
    registry: &std::sync::Arc<RwLock<StreamRegistry>>,
) -> (Child, JoinHandle<()>) {
    let mut child = Command::new(&args[0])
        .args(&args[1..])
        .args(["--stream-id", id])
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .unwrap_or_else(|e| panic!("failed to spawn {}: {e}", args[0]));

    if let Err(e) = job.assign(&child) {
        log::warn!("[{id}] failed to assign to job object: {e}");
    }

    let stdout = child.stdout.take().expect("stdout must be piped");

    let id_owned = id.to_owned();
    let registry_clone = Arc::clone(registry);

    // Stdout reader: blocking read_message() on a dedicated thread.
    // Uses blocking_write() to push frames — hold time is microseconds per
    // frame, so contention with HTTP readers is negligible.
    let reader_handle = tokio::task::spawn_blocking(move || {
        let mut reader = BufReader::new(stdout);
        loop {
            match live_video::read_message(&mut reader) {
                Ok(Some(msg)) => {
                    let mut registry = registry_clone.blocking_write();
                    if let Some(stream) = registry.streams.get_mut(&id_owned) {
                        match msg {
                            Message::CodecParams(params) => {
                                stream.buffer.set_codec_params(params);
                                if stream.status == StreamStatus::Starting {
                                    stream.status = StreamStatus::Running;
                                    log::info!("@{id_owned} running (codec params received)");
                                }
                            }
                            Message::Frame(frame) => {
                                stream.buffer.push_frame(&frame);
                            }
                            Message::Error(e) => {
                                log::error!("@{id_owned} capture error: {e}");
                            }
                        }
                    }
                    drop(registry);
                }
                Ok(None) => {
                    log::info!("@{id_owned} stdout EOF");
                    break;
                }
                Err(e) => {
                    log::error!("@{id_owned} read error: {e}");
                    break;
                }
            }
        }

        // Mark stream as stopped.
        let mut registry = registry_clone.blocking_write();
        if let Some(stream) = registry.streams.get_mut(&id_owned) {
            stream.status = StreamStatus::Stopped;
        }
    });

    (child, reader_handle)
}

/// Generate a short random ID (8 hex chars).
fn short_id() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let t = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default();
    format!("{:08x}", t.as_nanos() as u32)
}
